### ISSUE TITLE
MUI inputRef type error

### ISSUE THREAD

**Issue by cavator** [uNmAnNeR/imaskjs#639]
i'm folllowing the example on MUI doc to use imask
[MUI_Docs](https://mui.com/components/text-fields/#integration-with-3rd-party-input-libraries)
[codeSandbox](https://codesandbox.io/s/81c15f?file=/demo.tsx)

but when i try to just copy and past this solution i face numerous issues
like   
inputRef type error
TS2322: Type 'ForwardedRef<HTMLElement>' is not assignable to type '((instance: ReactElement | null) => void) | undefined'.   Type 'null' is not assignable to type '((instance: ReactElement | null) => void) | undefined'.
and when i try to update the value, it goes in loop 
Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.

The above error occurred in the <ForwardRef(InputBase)> component:

can someone test it outside the codesandox, to give me some answers, in the codesandbox it doesn't accuse the type errors

```TypeScript
import React from 'react';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import TextField from '@mui/material/TextField';
import Select from '@mui/material/Select';
import Toolbar from '@mui/material/Toolbar';
import Button from '@mui/material/Button';
import Divider from '@mui/material/Divider';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import { IMaskInput } from 'react-imask';
import Input from '@mui/material/Input';

interface CnpjProps {
  // eslint-disable-next-line no-unused-vars
  onChange: (event: { target: { name: string; value: string } }) => void;
  name: string;
}

function Partner() {
  const [formValues, setFormValues] = React.useState({
    cnpj: '',
    age: ''
  });
  const handleChange = (event: any) => {
    console.log(event);
    console.log(event.target.value, event.target.name, formValues.cnpj);
    setFormValues({
      ...formValues,
      [event.target.name]: event.target.value
    });
  };

  const CnpjMask = React.forwardRef<HTMLElement, CnpjProps>(
    (props, ref) => {
      const { onChange, ...other } = props;
      return (
        <IMaskInput
          {...other}
          mask="00.000.000/0000-00"
          inputRef={ref}
          onAccept={(value: any) => onChange({ target: { name: props.name, value } })}
          overwrite
        />
      );
    }
  );
  return (
    <Box component="form" m="auto" flexDirection="column" alignItems="center">
      <Box
        component={Grid}
        m="auto"
        item
        display={{ xs: 'none', sm: 'block' }}
        flexDirection="column"
        alignItems="center"
        sx={{
          bgcolor: '#FFF',
          p: 3
        }}
        xs={8}
      >
        <Toolbar />
        <Grid container spacing={2}>
          <Grid item xs={8}>
            <TextField fullWidth autoFocus variant="outlined" label="Razao Social" />
          </Grid>
          <Grid item xs={4}>
            <TextField fullWidth variant="outlined" label="Nome Fantasia" />
          </Grid>
          <Grid item xs={3}>
            <FormControl variant="outlined">
              <InputLabel variant="outlined" htmlFor="cnpj-mask-input">CNPJ</InputLabel>
              <Input
                value={formValues.cnpj}
                onChange={handleChange}
                name="cnpj"
                id="cnpj-mask-input"
                inputComponent={CnpjMask as any}
              />
            </FormControl>
          </Grid>
          <Grid item xs={6}>
            <TextField fullWidth variant="outlined" label="Website" />
          </Grid>
          <Grid item xs={3}>
            <FormControl fullWidth>
              <InputLabel id="demo-simple-select-label">Age</InputLabel>
              <Select
                labelId="demo-simple-select-label"
                id="demo-simple-select"
                value={formValues.age}
                name="age"
                label="Age"
                onChange={handleChange}
              >
                <MenuItem value={10}>Ten</MenuItem>
                <MenuItem value={20}>Twenty</MenuItem>
                <MenuItem value={30}>Thirty</MenuItem>
              </Select>
            </FormControl>
          </Grid>
        </Grid>
        <Divider textAlign="left">Logradouro</Divider>
        <Grid container spacing={2}>
          <Grid item xs={8}>
            <TextField fullWidth variant="outlined" label="Endereco" />
          </Grid>
          <Grid item xs={4}>
            <TextField fullWidth variant="outlined" label="Bairro" />
          </Grid>
          <Grid item xs={4}>
            <TextField fullWidth variant="outlined" label="Cidade" />
          </Grid>
          <Grid item xs={2}>
            <TextField fullWidth variant="outlined" label="UF" />
          </Grid>
          <Grid item xs={4}>
            <TextField fullWidth variant="outlined" label="CEP" />
          </Grid>
          <Grid item xs={2}>
            <TextField fullWidth variant="outlined" label="Telefone" />
          </Grid>
        </Grid>
        <Button type="submit" color="success" variant="contained">
          Salvar
        </Button>
      </Box>
    </Box>
  );
}

export default Partner;

```

this is my code if anyone want to try, it's just a simple forrm
ps: the update for the select works perfectly, just the IMask is giving me issues

**Comment by farizsofyan** [comment]
Try to put CnpjMask outside the function (preferrable) or add useMemo to CnpjMask

**Comment by kvnngn** [comment]
I am facing the same issue than @cavator. Any update ?

`import React from 'react';
import { DateTimePicker, LocalizationProvider } from '@mui/lab';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import { Box, TextField } from '@mui/material';
import CalendarTodayIcon from '@mui/icons-material/CalendarToday';
import moment from 'moment';
import { IMaskInput, IMask } from 'react-imask';
import { GrmtDateTimeRangePickerProps } from './datetimerange.types';

interface CustomProps {
  onChange: (event: { target: { name: string; value: string } }) => void;
  name: string;
}

const DatetimeMask = React.forwardRef<HTMLElement, CustomProps>(
  (props, ref) => {
    const { onChange, ...other } = props;
    const blocks = {
      YYYY: {
        mask: IMask.MaskedRange,
        from: 1970,
        to: 2030,
      },
      MM: {
        mask: IMask.MaskedRange,
        from: 1,
        to: 12,
      },
      DD: {
        mask: IMask.MaskedRange,
        from: 1,
        to: 31,
      },
      HH: {
        mask: IMask.MaskedRange,
        from: 0,
        to: 23,
      },
      mm: {
        mask: IMask.MaskedRange,
        from: 0,
        to: 59,
      },
    };

    return (
      <IMaskInput
        {...other}
        mask={Date}
        blocks={blocks}
        pattern="YYYY/MM/DD HH:mm"
        inputRef={ref}
        onAccept={(value: any) => {
          if (value && props.name) {
            onChange({ target: { name: props.name, value } });
          }
        }}
        overwrite
      />
    );
  },
);

const GrmtDateTimeRangePicker: React.FC<GrmtDateTimeRangePickerProps> = ({
  startDate,
  endDate,
  setEndDate,
  setStartDate,
}) => (
  <>
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DateTimePicker
        components={{ OpenPickerIcon: CalendarTodayIcon }}
        hideTabs
        inputFormat="yyyy/MM/dd HH:mm"
        PopperProps={{
          placement: 'bottom-end',
          disablePortal: true,
        }}
        ampm={false}
        maxDateTime={moment().toDate()}
        value={startDate}
        onChange={(newValue) => {
          const format = moment(newValue).format('YYYY-MM-DDTHH:mm:ss');
          setStartDate(format);
          if (
            !endDate ||
            (endDate && newValue > new Date(endDate)) ||
            new Date(endDate) > new Date()
          ) {
            setEndDate(format);
          }
        }}
        renderInput={(params) => (
          <TextField
            {...params}
            label="From"
            InputProps={{
              ...params.InputProps,
              inputComponent: DatetimeMask as any,
              style: {
                fontSize: '12px',
                color: '#737480',
                width: '165px',
                border: '1px solid #F0F0F7',
              },
            }}
          />
        )}
      />
    </LocalizationProvider>
    <Box mx={1}>{' ~ '}</Box>
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DateTimePicker
        components={{ OpenPickerIcon: CalendarTodayIcon }}
        inputFormat="yyyy/MM/dd HH:mm"
        PopperProps={{
          placement: 'bottom-end',
          disablePortal: true,
        }}
        ampm={false}
        minDateTime={new Date(startDate)}
        maxDateTime={new Date()}
        value={new Date(endDate)}
        onChange={(newValue) => {
          setEndDate(moment(newValue).format('YYYY-MM-DDTHH:mm:ss'));
        }}
        renderInput={(params) => (
          <TextField
            {...params}
            label="To"
            InputProps={{
              ...params.InputProps,
              inputComponent: DatetimeMask as any,
              style: {
                fontSize: '12px',
                color: '#737480',
                width: '165px',
                border: '1px solid #F0F0F7',
              },
            }}
          />
        )}
      />
    </LocalizationProvider>
  </>
);

export default GrmtDateTimeRangePicker;
`

**Comment by romeugodoi** [comment]
The same here

**Comment by KiwiKilian** [comment]
I'm currently using the following workaround:

```tsx
inputRef={ref as RefCallback<HTMLTextAreaElement | HTMLInputElement>}
```
Looks like a problem of the typing of `inputRef` when combined with `forwardRef`? Except the currently broken docs of MUI it doesn't seem to be related with it.

If anyone has a cleaner solution for using `forwardRef` with `IMaskInput` please drop a hint!

**Comment by pzi** [comment]
FWIW, the example in the MUI docs work because they are actively overriding (augmenting) `react-imask`s types here https://github.com/mui/material-ui/pull/27071/files/4b1f549d66bb18b45f8ef3c378224ad73dd5b5d2#diff-4f06b9ada6580c5f3205b95472832bab5e719bba3aa35433a5d00272db3bca24

So you can either augment the types yourself so `inputRef` does accept whatever `ref` is (`React.ForwardedRef<HTMLInputElement>`) or you cast it every time you use it.

**Comment by uNmAnNeR** [comment]
fixed in #667 
types will be refactored in 7.0.0
